extensions/rich_text_components/rte-output-display.component.ts

```diff
@@ -0,0 +1,178 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Component for the showing rich text.
+ */
+
+import { TemplatePortal } from '@angular/cdk/portal';
+import { AfterViewInit, ChangeDetectorRef, Component, Directive, ElementRef, Input, SimpleChanges, TemplateRef, ViewChild, ViewContainerRef } from '@angular/core';
+import { downgradeComponent } from '@angular/upgrade/static';
+import { OppiaRteParserService, OppiaRteNode, TextNode } from 'services/oppia-rte-parser.service';
+
+type PortalTree = (TemplatePortal<unknown> | PortalTree) [];
+
+@Component({
+  selector: 'oppia-rte-output-display',
+  templateUrl: './rte-output-display.component.html',
+  styleUrls: []
+})
+export class RteOutputDisplayComponent implements AfterViewInit {
+  // Native HTML elements.
+  @ViewChild('p') pTagPortal: TemplateRef<unknown>;
+  @ViewChild('span') spanTagPortal: TemplateRef<unknown>;
+  @ViewChild('ol') olTagPortal: TemplateRef<unknown>;
+  @ViewChild('li') liTagPortal: TemplateRef<unknown>;
+  @ViewChild('ul') ulTagPortal: TemplateRef<unknown>;
+  @ViewChild('pre') preTagPortal: TemplateRef<unknown>;
+  @ViewChild('strong') strongTagPortal: TemplateRef<unknown>;
+  @ViewChild('blockquote') blockquoteTagPortal: TemplateRef<unknown>;
+  @ViewChild('em') emTagPortal: TemplateRef<unknown>;
+  @ViewChild('text') textTagPortal: TemplateRef<unknown>;
+  // Oppia Non interactive.
+  @ViewChild('collapsible') collapsibleTagPortal: TemplateRef<unknown>;
+  @ViewChild('image') imageTagPortal: TemplateRef<unknown>;
+  @ViewChild('link') linkTagPortal: TemplateRef<unknown>;
+  @ViewChild('math') mathTagPortal: TemplateRef<unknown>;
+  @ViewChild('skillreview') skillreviewTagPortal: TemplateRef<unknown>;
+  @ViewChild('svgdiagram') svgdiagramTagPortal: TemplateRef<unknown>;
+  @ViewChild('tabs') tabsTagPortal: TemplateRef<unknown>;
+  @ViewChild('video') videoTagPortal: TemplateRef<unknown>;
+  @Input() rteString: string;
+  node: OppiaRteNode | string = '';
+  show = false;
+  portalTree: PortalTree = [];
+
+  constructor(
+    private _viewContainerRef: ViewContainerRef,
+    private cdRef: ChangeDetectorRef,
+    private elementRef: ElementRef,
+    private oppiaHtmlParserService: OppiaRteParserService
+  ) {}
+
+  private _updateNode(): void {
+    if (this.rteString === undefined || this.rteString === null) {
+      return;
+    }
+    // When there are trailing spaces in the HTML, CKEditor adds &nbsp;
+    // to the HTML (eg: '<p> Text &nbsp; &nbsp; %nbsp;</p>'), which can
+    // lead to UI issues when displaying it. Hence, the following block
+    // replaces the trailing ' &nbsp; &nbsp; %nbsp;</p>' with just '</p>'.
+    // We can't just find and replace '&nbsp;' here since, those in the
+    // middle may actually be required. Only the trailing ones need to be
+    // replaced.
+    this.rteString = this.rteString.replace(/(&nbsp;(\s)?)*(<\/p>)/g, '</p>');
```

Comment:
Can this be done more semantically (find p's and look at their textContent, then modify those) rather than using string manipulation? Using string manipulation on HTML is very much an antipattern, in general.

Reply:
Took these from angular html bind. But the way this component is built, it that you can't just show anything because of a very strict allowlist of content/ tags. Other are just ignored and don't make their way into the html. As you can see in the lines above (https://github.com/oppia/oppia/pull/13229/files#diff-7ab0a76c85f4cd0e3763a5c62d33d28a1562845c54556eceefc1c2e22bbf9e1cR33-R51), there is a whitelist and no other tags are allowed. These string manipulations happen before the it's passed through the allowlist filter. There is only one change that happens after the rendering and thats in the setTimeout function below (https://github.com/oppia/oppia/pull/13229/files#diff-7ab0a76c85f4cd0e3763a5c62d33d28a1562845c54556eceefc1c2e22bbf9e1cR97-R123). That operation is very safely handled.